File: backends/node/shared/handler/src/base-handler.ts
import {
  trace,
  context as otelContext,
  metrics,
  SpanStatusCode,
  type Tracer,
  type Meter,
  type Counter,
  type Histogram,
  type Span,
} from "@opentelemetry/api";
import { D2Result, type InputError } from "@d2/result";
import type { ZodType, ZodError } from "zod";
import type { IHandlerContext } from "./i-handler-context.js";
import type { IHandler } from "./i-handler.js";
import { DEFAULT_HANDLER_OPTIONS, type HandlerOptions } from "./handler-options.js";
import type { RedactionSpec } from "./redaction-spec.js";

/** OTel instruments shared across all handler instances. */
interface HandlerInstruments {
  duration: Histogram;
  invocations: Counter;
  failures: Counter;
  exceptions: Counter;
}

/**
 * Base handler providing automatic logging, tracing, and metrics.
 * Mirrors D2.Shared.Handler.BaseHandler<THandler, TInput, TOutput> in .NET.
 *
 * Subclasses implement `executeAsync()` with the handler's core logic.
 */
export abstract class BaseHandler<TInput, TOutput> implements IHandler<TInput, TOutput> {
  // Static OTel instruments (shared across all handler instances, lazy-initialized)
  private static tracer: Tracer;
  private static meter: Meter;
  private static instruments: HandlerInstruments | undefined;

  protected readonly context: IHandlerContext;

  protected get traceId(): string | undefined {
    return (
      this.context.request.traceId ?? trace.getSpan(otelContext.active())?.spanContext().traceId
    );
  }

  /** Override to declare this handler's redaction posture. */
  get redaction(): RedactionSpec | undefined {
    return undefined;
  }

  /**
   * Returns the effective request context for telemetry enrichment.
   *
   * By default, returns `this.context.request` (the DI-scoped context).
   * Pre-auth singleton handlers (e.g., rate limit, throttle) should override
   * this to return the per-request context from their input, since their
   * DI context is the static service-level default.
   */
  protected getEffectiveRequest(_input: TInput): import("./i-request-context.js").IRequestContext {
    return this.context.request;
  }

  protected constructor(context: IHandlerContext) {
    this.context = context;

    // Lazy-init static OTel instruments on first instance creation
    if (!BaseHandler.instruments) {
      BaseHandler.tracer = trace.getTracer("d2.shared.handler");
      BaseHandler.meter = metrics.getMeter("d2.shared.handler");
      BaseHandler.instruments = {
        duration: BaseHandler.meter.createHistogram("d2.handler.duration", {
          description: "Handler execution duration in milliseconds",
          unit: "ms",
        }),
        invocations: BaseHandler.meter.createCounter("d2.handler.invocations", {
          description: "Total number of handler invocations",
        }),
        failures: BaseHandler.meter.createCounter("d2.handler.failures", {
          description: "Total number of handler failures (non-success results)",
        }),
        exceptions: BaseHandler.meter.createCounter("d2.handler.exceptions", {
          description: "Total number of unhandled exceptions in handlers",
        }),
      };
    }
  }

  async handleAsync(
    input: TInput,
    options?: HandlerOptions,
  ): Promise<D2Result<TOutput | undefined>> {
    const handlerName = this.constructor.name;
    const opts = { ...DEFAULT_HANDLER_OPTIONS, ...options };
    const attrs = { "handler.name": handlerName };

    // Record invocation
    BaseHandler.instruments!.invocations.add(1, attrs);

    // Use the effective request context — for scoped handlers this is
    // this.context.request (DI-populated); for pre-auth singletons the
    // override returns input.requestContext (the per-request context).
    const effectiveReq = this.getEffectiveRequest(input);

    // Enrich the parent span (e.g., gRPC server span, HTTP span) with context
    // attributes so they appear on the top-level span in Tempo — not just on
    // this handler's child span. Harmless if middleware already set them.
    // For Hono services, trace.getActiveSpan() may return null due to the
    // async context issue — the Hono middleware handles enrichment instead.
    BaseHandler.enrichRequestSpan(trace.getActiveSpan(), effectiveReq);

    return BaseHandler.tracer.startActiveSpan(handlerName, async (span) => {
      // Set handler metadata + request context on child span.
      span.setAttribute("handler.type", handlerName);
      span.setAttribute("trace.id", this.traceId ?? "");
      BaseHandler.setIfPresent(span, "user.id", effectiveReq.userId);
      BaseHandler.setIfPresent(span, "agent.org.id", effectiveReq.agentOrgId);
      BaseHandler.setIfPresent(span, "target.org.id", effectiveReq.targetOrgId);

      const startTime = performance.now();

      try {
        // Log execution start
        this.context.logger.info(`Executing handler ${handlerName}. TraceId: ${this.traceId}.`);

        // Log input as debug if enabled (respecting redaction)
        if (opts.logInput && !this.redaction?.suppressInput) {
          const inputToLog = this.redaction?.inputFields?.length
            ? this.redactForLogging(input, this.redaction.inputFields)
            : input;
          this.context.logger.debug(
            `Handler ${handlerName} received input: ${BaseHandler.safeStringify(inputToLog)}. TraceId: ${this.traceId}.`,
          );
        }

        // Execute the handler's logic
        const rawResult = await this.executeAsync(input);

        // Auto-inject ambient traceId into result if handler didn't set it.
        // This eliminates the need for every handler to pass `traceId: this.traceId`.
        const result = this._injectTraceId(rawResult);

        // Stop timing
        const elapsedMs = performance.now() - startTime;

        // Log output as debug if enabled (respecting redaction)
        if (opts.logOutput && !this.redaction?.suppressOutput) {
          const resultToLog =
            this.redaction?.outputFields?.length && result.data
              ? { ...result, data: this.redactForLogging(result.data, this.redaction.outputFields) }
              : result;
          this.context.logger.debug(
            `Handler ${handlerName} produced result: ${BaseHandler.safeStringify(resultToLog)}. TraceId: ${this.traceId}.`,
          );
        }

        // Set result metadata on span
        span.setAttribute("handler.success", result.success);
        span.setAttribute("handler.status.code", result.statusCode);
        span.setAttribute("handler.error.code", result.errorCode ?? "");
        span.setAttribute("handler.elapsed.ms", elapsedMs);
        span.setStatus(
          result.success ? { code: SpanStatusCode.OK } : { code: SpanStatusCode.ERROR },
        );

        // Record duration metric
        BaseHandler.instruments!.duration.record(elapsedMs, attrs);

        // Record failure metric if not successful
        if (!result.success) {
          BaseHandler.instruments!.failures.add(1, attrs);
        }

        // Determine log level based on success and elapsed time
        const status = result.success ? "successfully" : "unsuccessfully";
        const completionMsg = `Executed handler ${handlerName} ${status} in ${elapsedMs.toFixed(0)}ms. TraceId: ${this.traceId}.`;

        if (opts.suppressTimeWarnings) {
          if (result.success) {
            this.context.logger.info(completionMsg);
          } else {
            this.context.logger.warn(completionMsg);
          }
        } else if (result.success && elapsedMs < opts.warningThresholdMs) {
          this.context.logger.info(completionMsg);
        } else if (elapsedMs < opts.criticalThresholdMs) {
          this.context.logger.warn(completionMsg);
        } else {
          this.context.logger.error(completionMsg);
        }

        span.end();
        return result;
      } catch (ex) {
        // Stop timing
        const elapsedMs = performance.now() - startTime;

        // Create unhandled exception result
        const errorResult = D2Result.unhandledException<TOutput | undefined>({
          traceId: this.traceId,
        });

        // Set exception metadata on span
        span.setAttribute("handler.success", errorResult.success);
        span.setAttribute("handler.status.code", errorResult.statusCode);
        span.setAttribute("handler.error.code", errorResult.errorCode ?? "");
        span.setAttribute("handler.elapsed.ms", elapsedMs);
        span.setStatus({ code: SpanStatusCode.ERROR });
        if (ex instanceof Error) {
          span.recordException(ex);
        }

        // Record metrics
        BaseHandler.instruments!.duration.record(elapsedMs, attrs);
        BaseHandler.instruments!.exceptions.add(1, attrs);

        // Log the unhandled exception
        this.context.logger.error(
          `Handler ${handlerName} encountered an unhandled exception after ${elapsedMs.toFixed(0)}ms. TraceId: ${this.traceId}.`,
        );

        span.end();
        return errorResult;
      }
    });
  }

  /**
   * Executes the core logic of the handler.
   * Subclasses must implement this method.
   */
  protected abstract executeAsync(input: TInput): Promise<D2Result<TOutput | undefined>>;

  /**
   * Validates the input against a Zod schema.
   * Returns ok on success, validationFailed with inputErrors on failure.
   */
  protected validateInput(schema: ZodType<TInput>, input: TInput): D2Result<void> {
    const result = schema.safeParse(input);
    if (result.success) {
      return D2Result.ok();
    }

    const inputErrors: InputError[] = (result as { error: ZodError }).error.issues.map((issue) => [
      issue.path.join("."),
      issue.message,
    ]);
    return D2Result.validationFailed({ inputErrors });
  }

  /**
   * Injects the ambient traceId into a result if the handler didn't set one.
   * This enables handlers to omit `traceId: this.traceId` from every D2Result call.
   */
  private _injectTraceId<T>(result: D2Result<T>): D2Result<T> {
    if (result.traceId || !this.traceId) return result;
    return new D2Result({
      success: result.success,
      data: result.data,
      messages: [...result.messages],
      inputErrors: result.inputErrors.map((ie) => [...ie]),
      statusCode: result.statusCode,
      errorCode: result.errorCode,
      traceId: this.traceId,
    });
  }

  /** Safely serializes a value to JSON with size limiting and circular reference protection. */
  private static safeStringify(value: unknown, maxLength = 10_000): string {
    try {
      const json = JSON.stringify(value);
      if (json.length > maxLength) {
        return json.slice(0, maxLength) + `...[truncated, ${json.length} total chars]`;
      }
      return json;
    } catch {
      return "[unserializable]";
    }
  }

  /** Replaces specified top-level fields with "[REDACTED]" for logging. */
  private redactForLogging(obj: unknown, fields: readonly string[]): unknown {
    if (!fields.length || obj === null || obj === undefined || typeof obj !== "object") {
      return obj;
    }
    const clone: Record<string, unknown> = { ...(obj as Record<string, unknown>) };
    for (const field of fields) {
      if (field in clone) {
        clone[field] = "[REDACTED]";
      }
    }
    return clone;
  }

  /** Sets a span attribute only if the value is non-null/undefined. */
  private static setIfPresent(span: Span, key: string, value: string | undefined): void {
    if (value != null) span.setAttribute(key, value);
  }

  /**
   * Enriches a span with standard request context attributes.
   * Used on both parent spans (top-level HTTP/gRPC) and child handler spans.
   */
  private static enrichRequestSpan(
    span: Span | undefined,
    req: import("./i-request-context.js").IRequestContext,
  ): void {
    if (!span) return;
    // String attributes — only set if present.
    for (const [key, value] of Object.entries({
      userId: req.userId,
      username: req.username,
      agentOrgId: req.agentOrgId,
      agentOrgType: req.agentOrgType != null ? String(req.agentOrgType) : undefined,
      agentOrgRole: req.agentOrgRole,
      targetOrgId: req.targetOrgId,
      targetOrgType: req.targetOrgType != null ? String(req.targetOrgType) : undefined,
    })) {
      if (value != null) span.setAttribute(key, value);
    }
    // Boolean flags — always set (default false).
    span.setAttribute("isAuthenticated", req.isAuthenticated ?? false);
    span.setAttribute("isTrustedService", req.isTrustedService ?? false);
    if (req.isAuthenticated) {
      span.setAttribute("isOrgEmulating", req.isOrgEmulating ?? false);
    }
  }
}

File: backends/dotnet/shared/Handler/BaseHandler.cs
// -----------------------------------------------------------------------
// <copyright file="BaseHandler.cs" company="DCSV">
// Copyright (c) DCSV. All rights reserved.
// </copyright>
// -----------------------------------------------------------------------

namespace D2.Shared.Handler;

using System.Diagnostics;
using D2.Shared.Result;
using FluentValidation;
using Microsoft.Extensions.Logging;

/// <summary>
/// A base implementation of the <see cref="IHandler{TInput, TOutput}"/> interface that provides
/// common functionality for handling requests, including logging, timing, and exception
/// handling.
/// </summary>
///
/// <typeparam name="THandler">
/// The type of the handler's implementation.
/// </typeparam>
/// <typeparam name="TInput">
/// The type of input the handler processes.
/// </typeparam>
/// <typeparam name="TOutput">
/// The type of output the handler produces.
/// </typeparam>
public abstract class BaseHandler<THandler, TInput, TOutput> : IHandler<TInput, TOutput>
{
    /// <summary>
    /// Initializes a new instance of the <see cref="BaseHandler{THandler, TInput, TOutput}"/>
    /// class.
    /// </summary>
    ///
    /// <param name="context">
    /// The handler context for the current operation.
    /// </param>
    protected BaseHandler(IHandlerContext context)
    {
        Context = context;
    }

    /// <summary>
    /// Gets the handler context for the current operation.
    /// </summary>
    protected IHandlerContext Context { get; }

    /// <summary>
    /// Gets the trace ID associated with the current request.
    /// </summary>
    protected string? TraceId => Context.Request.TraceId;

    /// <summary>
    /// Gets the default handler options for this handler type.
    /// Override to customize logging behavior (e.g., suppress input/output for sensitive data).
    /// </summary>
    protected virtual HandlerOptions DefaultOptions => new();

    /// <inheritdoc />
    /// <remarks>
    /// This method wraps the execution of the handler with logging, timing, and exception
    /// handling. It logs the start and end of the handler execution, measures the elapsed time,
    /// and logs any unhandled exceptions that occur during execution. The actual handler logic is
    /// implemented in the abstract <see cref="ExecuteAsync"/> method, which must be overridden by
    /// derived classes.
    /// </remarks>
    public async ValueTask<D2Result<TOutput?>> HandleAsync(
        TInput input,
        CancellationToken ct = default,
        HandlerOptions? options = null)
    {
        // Create a new activity for tracing.
        using var activity = BHASW.SR_ActivitySource.StartActivity(typeof(THandler).Name);

        // Add metadata to the activity.
        activity?.SetTag("handler.type", typeof(THandler).FullName ?? typeof(THandler).Name);
        activity?.SetTag("trace.id", Context.Request.TraceId);
        activity?.SetTag("user.id", Context.Request.UserId);
        activity?.SetTag("agent.org.id", Context.Request.AgentOrgId);
        activity?.SetTag("target.org.id", Context.Request.TargetOrgId);

        // Start the stopwatch to measure elapsed time.
        var sw = Stopwatch.StartNew();

        // Use default options if none are provided.
        options ??= DefaultOptions;

        // Record invocation metric.
        var tags = new TagList { { "handler.name", typeof(THandler).Name } };
        BHASW.SR_Invocations.Add(1, tags);

        // Wrap the execution in a try-catch to handle exceptions.
        try
        {
            // Log that we are now executing the handler.
            Context.Logger.LogInformation(
                "Executing handler {HandlerName}. TraceId: {TraceId}.",
                typeof(THandler).Name,
                Context.Request.TraceId);

            // Log the input as debug if enabled.
            if (options.LogInput)
            {
                Context.Logger.LogDebug(
                    "Handler {HandlerName} received input: {@Input}. TraceId: {TraceId}.",
                    typeof(THandler).Name,
                    input,
                    Context.Request.TraceId);
            }

            // Execute the handler's logic.
            var result = await ExecuteAsync(input, ct);

            // Auto-inject traceId if the handler didn't set it explicitly.
            result = InjectTraceId(result);

            // Stop the stopwatch to measure elapsed time.
            sw.Stop();
            var elapsedMs = sw.ElapsedMilliseconds;

            // Log the output as debug if enabled.
            if (options.LogOutput)
            {
                Context.Logger.LogDebug(
                    "Handler {HandlerName} produced result: {@Output}. TraceId: {TraceId}.",
                    typeof(THandler).Name,
                    result,
                    Context.Request.TraceId);
            }

            // Add result metadata to the activity.
            activity?.SetTag("handler.success", result.Success);
            activity?.SetTag("handler.status.code", result.StatusCode);
            activity?.SetTag("handler.error.code", result.ErrorCode);
            activity?.SetTag("handler.elapsed.ms", elapsedMs);
            activity?.SetStatus(result.Success ? ActivityStatusCode.Ok : ActivityStatusCode.Error);

            // Record duration metric and optional failure metric.
            BHASW.SR_Duration.Record(elapsedMs, tags);
            if (!result.Success)
            {
                BHASW.SR_Failures.Add(1, tags);
            }

            // Determine the log level based on success and elapsed time.
            var level = options.SuppressTimeWarnings
                ? result.Success
                    ? LogLevel.Information
                    : LogLevel.Warning
                : result.Success && elapsedMs < options.WarningThresholdMs
                    ? LogLevel.Information
                    : elapsedMs < options.CriticalThresholdMs
                        ? LogLevel.Warning
                        : LogLevel.Critical;

            // Log the completion of the handler execution.
            Context.Logger.Log(
                level,
                "Executed handler {HandlerName} {Status} in {ElapsedMilliseconds}ms. TraceId: {TraceId}.",
                typeof(THandler).Name,
                result.Success ? "successfully" : "unsuccessfully",
                elapsedMs,
                Context.Request.TraceId);

            // Return the result.
            return result;
        }
        catch (Exception ex)
        {
            // Stop the stopwatch.
            sw.Stop();
            var elapsedMs = sw.ElapsedMilliseconds;

            // Create an unhandled exception result.
            var errorResult = D2Result<TOutput?>.UnhandledException(
                traceId: Context.Request.TraceId);

            // Add exception metadata to the activity.
            activity?.SetTag("handler.success", errorResult.Success);
            activity?.SetTag("handler.status.code", errorResult.StatusCode);
            activity?.SetTag("handler.error.code", errorResult.ErrorCode);
            activity?.SetTag("handler.elapsed.ms", elapsedMs);
            activity?.SetStatus(ActivityStatusCode.Error);
            activity?.AddException(ex);

            // Record exception duration and exception counter metrics.
            BHASW.SR_Duration.Record(elapsedMs, tags);
            BHASW.SR_Exceptions.Add(1, tags);

            // Log the unhandled exception.
            Context.Logger.LogError(
                ex,
                "Handler {HandlerName} encountered an unhandled exception after {ElapsedMilliseconds}ms. TraceId: {TraceId}.",
                typeof(THandler).Name,
                elapsedMs,
                Context.Request.TraceId);

            // Return the error result.
            return errorResult;
        }
    }

    /// <summary>
    /// Executes the core logic of the handler.
    /// </summary>
    ///
    /// <param name="input">
    /// The input to be processed by the handler.
    /// </param>
    /// <param name="ct">
    /// A cancellation token to cancel the operation.
    /// </param>
    ///
    /// <returns>
    /// A <see cref="ValueTask{D2Result}"/> containing the result of the handler's execution.
    /// </returns>
    protected abstract ValueTask<D2Result<TOutput?>> ExecuteAsync(
        TInput input,
        CancellationToken ct = default);

    /// <summary>
    /// Validates the input using configurable <paramref name="options"/>.
    /// </summary>
    ///
    /// <param name="options">
    /// An optional action to configure the validator.
    /// </param>
    /// <param name="input">
    /// The input to be validated.
    /// </param>
    /// <param name="ct">
    /// A cancellation token to cancel the validation operation.
    /// </param>
    ///
    /// <returns>
    /// A <see cref="ValueTask{D2Result}"/> containing the result of the validation.
    /// </returns>
    protected async ValueTask<D2Result> ValidateInput(
        Action<AbstractValidator<TInput>>? options,
        TInput input,
        CancellationToken ct = default)
    {
        // Create the validator with the provided configuration.
        var validator = new ConfigurableValidator(options);

        // Validate the input asynchronously.
        var validationResult = await validator.ValidateAsync(input, ct);

        // If valid, return OK.
        if (validationResult.IsValid)
        {
            return D2Result.Ok();
        }

        // Group errors by property name and select into a 2D list.
        var errors = validationResult.Errors
            .GroupBy(e => e.PropertyName)
            .Select(g =>
                new List<string> { g.Key }
                    .Concat(g.Select(e => e.ErrorMessage))
                    .ToList())
            .ToList();

        // Return validation failed result with errors.
        return D2Result.ValidationFailed(
            inputErrors: errors);
    }

    /// <summary>
    /// Injects the ambient TraceId into the result if the handler didn't set it explicitly.
    /// This eliminates the need for every handler to pass <c>traceId: TraceId</c> manually.
    /// </summary>
    private D2Result<TOutput?> InjectTraceId(D2Result<TOutput?> result)
    {
        if (result.TraceId is not null || TraceId is null)
        {
            return result;
        }

        return new D2Result<TOutput?>(
            result.Success,
            result.Data,
            result.Messages,
            result.InputErrors,
            result.StatusCode,
            result.ErrorCode,
            TraceId);
    }

    /// <summary>
    /// A configurable validator that allows dynamic configuration of validation rules via an
    /// action.
    /// </summary>
    private class ConfigurableValidator : AbstractValidator<TInput>
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="ConfigurableValidator"/> class.
        /// </summary>
        ///
        /// <param name="configure">
        /// An optional action to configure the validator.
        /// </param>
        public ConfigurableValidator(Action<AbstractValidator<TInput>>? configure = null)
            => configure?.Invoke(this);
    }
}
